activerecord/lib/active_record/tasks/database_tasks.rb

```diff
@@ -96,11 +97,11 @@ def fixtures_path
       end
 
       def root
-        @root ||= Rails.root
+        @root ||= Rails.root if defined?(Rails.root)
       end
 
       def env
-        @env ||= Rails.env
+        @env ||= ActiveRecord::ConnectionHandling::DEFAULT_ENV.call
```

Comment:
This method takes care of the case where `Rails.env` is not defined and in fact we use it in a couple of different places e.g. 

* https://github.com/rails/rails/blob/7bc343b3a9ef7581e665a9f70f00d494f7a3b62c/activerecord/lib/active_record/database_configurations/database_config.rb#L33
* https://github.com/rails/rails/blob/7bc343b3a9ef7581e665a9f70f00d494f7a3b62c/activerecord/lib/active_record/migration.rb#L1131

It might make sense to move this out into a `ActiveRecord::Base.current_environment` method and have that be the canonical way of fetching the environment.

Reply:
EDIT: proposed another PR for this fix: https://github.com/rails/rails/pull/36435